boa_interner/src/lib.rs

```diff
@@ -69,46 +70,68 @@
     rustdoc::missing_doc_code_examples
 )]
 
+mod fixed_string;
+mod interned_str;
+mod sym;
 #[cfg(test)]
 mod tests;
 
-use std::{fmt::Display, num::NonZeroUsize};
+use fixed_string::FixedString;
+pub use sym::*;
 
-use gc::{unsafe_empty_trace, Finalize, Trace};
-#[cfg(feature = "serde")]
-use serde::{Deserialize, Serialize};
-use string_interner::{backend::BucketBackend, StringInterner, Symbol};
+use std::fmt::{Debug, Display};
 
-/// Backend of the string interner.
-type Backend = BucketBackend<Sym>;
+use interned_str::InternedStr;
+use rustc_hash::FxHashMap;
 
 /// The string interner for Boa.
-///
-/// This is a type alias that makes it easier to reference it in the code.
-#[derive(Debug)]
+#[derive(Debug, Default)]
 pub struct Interner {
-    inner: StringInterner<Backend>,
+    // COMMENT FOR DEVS:
+    // This interner works on the assumption that
+    // `head` won't ever be reallocated, since this could invalidate
+    // some of our stored pointers inside `spans`.
+    // This means that any operation on `head` and `full` should be carefully
+    // reviewed to not cause Undefined Behaviour.
+    // `get_or_intern` has a more thorough explanation on this.
+    //
+    // Also, if you want to implement `shrink_to_fit` (and friends),
+    // please check out https://github.com/Robbepop/string-interner/pull/47 first.
+    // This doesn't implement that method, since implementing it increases
+    // our memory footprint.
+    symbols: FxHashMap<InternedStr, Sym>,
+    spans: Vec<InternedStr>,
+    head: FixedString,
+    full: Vec<FixedString>,
 }
 
 impl Interner {
-    /// Creates a new `StringInterner` with the given initial capacity.
+    /// Creates a new [`Interner`].
     #[inline]
-    pub fn with_capacity(cap: usize) -> Self {
+    pub fn new() -> Self {
+        Self::default()
+    }
```

Comment:
Do we need a `new()` function if we already have a `default()` function? This is usually frowned upon by clippy, for example.

Reply:
> This is usually frowned upon by clippy.

Not at all! On the contrary, Clippy recommends to have a `Default` implementation when you have a `new()` method.

https://rust-lang.github.io/rust-clippy/master/index.html#new_without_default

Reply:
Oh, indeed, I thought there was another lint that preferred the usage of `default()` over `new()` (I think in the "restriction" or "pedantic" groups), maybe it was removed, or maybe I'm just remembering it wrong xD but it's fine for me :)